crates/rethnet_evm/src/block/builder.rs

```diff
@@ -151,30 +236,101 @@ where
 
         self.header.gas_used += U256::from(result.gas_used());
 
-        // TODO: store receipt
+        let logs: Vec<Log> = result.logs().into_iter().map(Log::from).collect();
+        let logs_bloom = {
+            let mut bloom = Bloom::zero();
+            logs.iter().for_each(|log| {
+                log.add_to_bloom(&mut bloom);
+            });
+            bloom
+        };
+
+        let receipt = EIP658Receipt {
+            status_code: if result.is_success() { 1 } else { 0 },
+            gas_used: self.header.gas_used,
+            logs_bloom,
+            logs,
+        };
+
+        self.receipts.push(match &transaction.transaction {
+            SignedTransaction::Legacy(_) => TypedReceipt::Legacy(receipt),
+            SignedTransaction::EIP2930(_) => TypedReceipt::EIP2930(receipt),
+            SignedTransaction::EIP1559(_) => TypedReceipt::EIP1559(receipt),
+        });
+
+        let (transaction, caller) = transaction.into_inner();
+
+        self.callers.push(caller);
+        self.transactions.push(transaction);
+
         Ok(result)
     }
 
-    /// Finalizes the block, returning the state root.
-    /// TODO: Build a full block
-    pub async fn finalize(self, rewards: Vec<(Address, U256)>) -> Result<(), SE> {
+    /// Finalizes the block, returning the block and the callers of the transactions.
+    pub async fn finalize(
+        mut self,
```

Comment:
Just checking my understanding here: this BlockBuilder doesn't need to keep track of the started/sealed/reverted state because `finalize` and `abort` consume `self`, right?

Reply:
Exactly!